x/ibc/src/types/context/context.rs

```diff
@@ -0,0 +1,639 @@
+#[allow(dead_code)] //TODO: remove this when ready
+#[allow(unused_variables)] // TODO: remove
+#[allow(unused_imports)] //TODO: remove
+use derive_more::{From, TryInto};
+use gears::params::ParamsSubspaceKey;
+use gears::store::database::Database;
+use gears::store::QueryableKVStore;
+use gears::store::StoreKey;
+use gears::types::context::tx::TxContext;
+use ibc::clients::tendermint::client_state::ClientState as TmClientState;
+use ibc::clients::tendermint::consensus_state::ConsensusState as TmConsensusState;
+use ibc::clients::tendermint::types::{
+    ClientState as ClientStateType, ConsensusState as ConsensusStateType,
+    TENDERMINT_CLIENT_STATE_TYPE_URL, TENDERMINT_CONSENSUS_STATE_TYPE_URL,
+};
+//use ibc::core::client::context::client_state::ClientStateValidation;
+use gears::store::TransactionalKVStore;
+use gears::store::WritePrefixStore;
+use gears::types::context::QueryableContext;
+use gears::types::context::TransactionalContext;
+use ibc::core::client::context::{
+    ClientExecutionContext, ClientValidationContext, ExtClientValidationContext,
+};
+use ibc::core::client::types::error::ClientError;
+use ibc::core::client::types::Height;
+use ibc::core::handler::types::error::ContextError;
+use ibc::core::host::types::identifiers::ClientId;
+use ibc::core::host::{ExecutionContext, ValidationContext};
+use ibc::core::primitives::proto::Protobuf;
+use ibc::core::router::router::Router;
+use ibc::derive::{ClientState, ConsensusState};
+use ibc::primitives::proto::Any;
+use ibc::primitives::Timestamp;
+use ibc::primitives::ToVec;
+use serde::Serialize;
+
+use crate::ics02_client::types::client_state::ClientState;
+use crate::ics02_client::types::consensus_state::ConsensusState;
+use crate::{
+    ics02_client::{message::MsgCreateClient, Keeper as ClientKeeper},
+    ics03_connection::Keeper as ConnectionKeeper,
+    ics04_channel::Keeper as ChannelKeeper,
+    params::IBCParamsKeeper,
+    types::genesis::GenesisState,
+};
+
+use crate::ics02_client::ClientParamsKeeper;
+//use super::keeper::KEY_NEXT_CLIENT_SEQUENCE;
+//use super::params::ClientParamsKeeper;
+use crate::ics02_client::KEY_NEXT_CLIENT_SEQUENCE;
+use crate::params::CLIENT_STATE_KEY;
+
+pub const KEY_CLIENT_STORE_PREFIX: &str = "clients";
+pub const KEY_CONSENSUS_STATE_PREFIX: &str = "consensusStates";
+pub const KEY_PROCESSED_TIME: &str = "/processedTime";
+
+// KeyProcessedHeight is appended to consensus state key to store the processed height
+const KEY_PROCESSED_HEIGHT: &str = "/processedHeight";
+const KEY_ITERATE_CONSENSUS_STATE_PREFIX: &[u8; 22] = b"iterateConsensusStates";
+
+pub struct Context<'a, 'b, DB, SK, PSK> {
+    pub gears_ctx: &'a mut TxContext<'b, DB, SK>,
```

Comment:
What do you think. Is it better to pass owned version of ctx to xmods or use mutable reference?

Reply:
For this PR, I believe it has to be a PR. Once your changes have been merged then we could look at changing this.